File: jstressy-platform/jstressy-impl/jstressy-metrics-micrometer/src/main/java/com/github/timofeevda/jstressy/metrics/micrometer/internal/Activator.java
package com.github.timofeevda.jstressy.metrics.micrometer.internal;

import com.github.timofeevda.jstressy.api.config.ConfigurationService;
import com.github.timofeevda.jstressy.api.metrics.MetricsRegistryService;
import com.github.timofeevda.jstressy.api.vertx.VertxService;
import com.github.timofeevda.jstressy.metrics.micrometer.MicrometerMetricsRegistryService;
import com.github.timofeevda.jstressy.utils.ServiceObserver;
import io.reactivex.Observable;
import io.reactivex.Single;
import org.osgi.framework.BundleActivator;
import org.osgi.framework.BundleContext;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Hashtable;

public class Activator implements BundleActivator {
    private static Logger logger = LoggerFactory.getLogger(Activator.class);

    @Override
    public void start(BundleContext context) {
        logger.info("Starting metrics registry service activator");

        Single<VertxService> vertxService = ServiceObserver.observeService(VertxService.class.getName(), context);
        Single<ConfigurationService> configurationService = ServiceObserver.observeService(ConfigurationService.class.getName(), context);

        Observable.combineLatest(
                vertxService.toObservable(),
                configurationService.toObservable(),
                this::toMetricsRegistryService)
                .subscribe(metricsRegistryService -> {
                    logger.info("Registering metrics registry service");
                    context.registerService(MetricsRegistryService.class.getName(), metricsRegistryService, new Hashtable());
                });
    }

    @Override
    public void stop(BundleContext context) {

    }

    private MetricsRegistryService toMetricsRegistryService(VertxService vertxService, ConfigurationService configurationService) {
        MicrometerMetricsRegistryService metricsRegistryService = new MicrometerMetricsRegistryService();
        metricsRegistryService.publishMetrics(vertxService, configurationService);
        return metricsRegistryService;
    }

}

File: jstressy-platform/jstressy-impl/jstressy-metrics-micrometer/src/main/kotlin/com/github/timofeevda/jstressy/metrics/micrometer/internal/Activator.kt
package com.github.timofeevda.jstressy.metrics.micrometer.internal

import com.github.timofeevda.jstressy.api.config.ConfigurationService
import com.github.timofeevda.jstressy.api.metrics.MetricsRegistryService
import com.github.timofeevda.jstressy.api.vertx.VertxService
import com.github.timofeevda.jstressy.metrics.micrometer.MicrometerMetricsRegistryService
import com.github.timofeevda.jstressy.utils.StressyUtils.observeService
import com.github.timofeevda.jstressy.utils.StressyUtils.serviceAwaitTimeout
import org.osgi.framework.BundleActivator
import org.osgi.framework.BundleContext
import org.slf4j.LoggerFactory
import java.util.*
import java.util.concurrent.TimeUnit

class Activator : BundleActivator {

    companion object {
        private val logger = LoggerFactory.getLogger(Activator::class.java)
    }

    override fun start(context: BundleContext) {
        logger.info("Starting metrics registry service activator")

        val metricsRegistryService = MicrometerMetricsRegistryService()

        observeService<ConfigurationService>(ConfigurationService::class.java.name, context)
                .doOnSubscribe { logger.info("Metric Registry service subscribes on Configuration services") }
                .doOnSuccess { logger.info("Registering metrics registry service") }
                .timeout(serviceAwaitTimeout().toMilliseconds(), TimeUnit.MILLISECONDS)
                .map { configurationService -> metricsRegistryService.setConfigurationService(configurationService) }
                .subscribe(
                        { metricsRegistry ->
                            context.registerService(MetricsRegistryService::class.java.name, metricsRegistry, Hashtable<Any, Any>())
                        },
                        { throwable -> logger.error("Error registering metrics registry service", throwable) }
                )

        observeService<VertxService>(VertxService::class.java.name, context)
                .doOnSubscribe { logger.info("Metric Registry service subscribes on VertX services") }
                .doOnSuccess { logger.info("Initializing Metrics Registry") }
                .timeout(serviceAwaitTimeout().toMilliseconds(), TimeUnit.MILLISECONDS)
                .subscribe(
                        { vertxService -> metricsRegistryService.startServingMetrics(vertxService)},
                        {throwable -> logger.error("Error initializing metrics registry service", throwable)}
                )

    }

    override fun stop(context: BundleContext) {
        logger.info("Stopping metrics registry service")
    }

}
